src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,333 @@
+            config.q_sinsemilla1.enable(region, offset + row)?;
+            config.q_sinsemilla2.enable(region, offset + row)?;
+        }
+
+        // Message piece as K * piece.length bitstring
+        let bitstring: Option<Vec<bool>> = piece.field_elem().map(|value| {
+            value
+                .to_le_bits()
+                .into_iter()
+                .take(K * piece.num_words())
+                .collect()
+        });
+
+        let words: Option<Vec<u32>> = bitstring.map(|bitstring| {
+            bitstring
+                .chunks_exact(K)
+                .map(|word| lebs2ip_k(word))
+                .collect()
+        });
+
+        // Get (x_p, y_p) for each word. We precompute this here so that we can use `batch_normalize()`.
+        let generators_projective: Option<Vec<C::CurveExt>> = words
+            .clone()
+            .map(|words| words.iter().map(|word| get_s_by_idx::<C>(*word)).collect());
+        let generators: Option<Vec<(C::Base, C::Base)>> =
+            generators_projective.map(|generators_projective| {
+                let mut generators = vec![C::default(); generators_projective.len()];
+                C::Curve::batch_normalize(&generators_projective, &mut generators);
+                generators
+                    .iter()
+                    .map(|gen| {
+                        let point = gen.coordinates().unwrap();
+                        (*point.x(), *point.y())
+                    })
+                    .collect()
+            });
+
+        // Convert `words` from `Option<Vec<u32>>` to `Vec<Option<u32>>`
+        let words: Vec<Option<u32>> = if let Some(words) = words {
+            words.into_iter().map(Some).collect()
+        } else {
+            vec![None; piece.num_words()]
+        };
+
+        // Decompose message into `K`-bit pieces with a running sum `z`.
+        let zs = {
+            let mut zs = Vec::with_capacity(piece.num_words() + 1);
+
+            // Copy message and initialize running sum `z` to decompose message in-circuit
+            let cell = region.assign_advice(
+                || "copy message",
+                config.word,
+                offset,
+                || piece.field_elem().ok_or(Error::SynthesisError),
+            )?;
+            region.constrain_equal(&config.perm, piece.cell(), cell)?;
+            zs.push(CellValue::new(cell, piece.field_elem()));
+
+            // Assign cumulative sum such that
+            //          z_i = 2^10 * z_{i + 1} + m_{i + 1}
+            // => z_{i + 1} = (z_i - m_{i + 1}) / 2^10
+            //
+            // For a message m = m_1 + 2^K m_2 + ... + 2^{Kl} m_l}, initialize z_0 = m.
+            // We end up with z_{l} = 0.
+            let mut z = piece.field_elem();
+            let inv_2_k = C::Base::from_u64(1 << K).invert().unwrap(); // TODO: Constant
+            for (idx, word) in words.iter().enumerate() {
+                // z_next = (z_cur - m_next) / 2^K
```

Comment:
The indices are slightly confusing here because Sinsemilla words are one-indexed (as per the [protocol spec](https://zips.z.cash/protocol/nu5.pdf#concretesinsemillahash)). So `m_{i + 1}` actually refers to the word on the same row as `z_i`.

![image](https://user-images.githubusercontent.com/14323671/121020813-4f749780-c7d3-11eb-99a3-0ed322d34689.png)

```suggestion
                // z_{i + 1} = (z_i - m_{i + 1}) / 2^10
```